source-oracle/replication.go

```diff
@@ -625,46 +729,86 @@ const (
 	opRollback = 36
 )
 
-func (s *replicationStream) generateLogminerQuery(ctx context.Context) error {
-	if s.logminerStmt != nil {
-		s.logminerStmt.Close()
-	}
+type transaction struct {
+	XID      string
+	StartSCN int64
+}
 
+func (s *replicationStream) generateLogminerQuery(ctx context.Context, transactions []string) (*sql.Stmt, error) {
 	var tableObjectMapping = s.db.tableObjectMapping
 
 	var conditions []string
 	for _, mapping := range tableObjectMapping {
 		if !s.tableActive(mapping.streamID) {
-			logrus.WithField("streamID", mapping.streamID).Debug("logminer: removing disabled table from replication query")
 			continue
 		}
 		conditions = append(conditions, fmt.Sprintf("(DATA_OBJ# = %d AND DATA_OBJD# = %d)", mapping.objectID, mapping.dataObjectID))
 	}
 	var opCodes = []int{opInsert, opDelete, opUpdate, opStart, opCommit, opRollback}
 	var opPredicate = strings.Trim(strings.Join(strings.Fields(fmt.Sprint(opCodes)), ","), "[]")
+
+	var txValues []string
+	for _, tx := range transactions {
+		txValues = append(txValues, fmt.Sprintf("utl_encode.base64_decode(utl_raw.cast_to_raw('%s'))", tx))
+	}
+	var txPredicate = ""
+	if len(txValues) > 0 {
+		txPredicate = fmt.Sprintf("AND XID IN (%s)", strings.Join(txValues, ","))
+	}
+
 	// START and COMMIT operations do not seg_owner and have data_obj#=0 and data_objd#=0
 	var query = fmt.Sprintf(`SELECT SCN, TIMESTAMP, OPERATION_CODE, SQL_REDO, SQL_UNDO, TABLE_NAME, SEG_OWNER, STATUS, INFO, RS_ID, SSN, CSF, DATA_OBJ#, DATA_OBJD#, XID
     FROM V$LOGMNR_CONTENTS
-    WHERE OPERATION_CODE IN (%s) AND SCN >= :startSCN AND SCN <= :endSCN AND
+    WHERE OPERATION_CODE IN (%s) AND SCN >= :startSCN AND SCN <= :endSCN AND 
     (SEG_OWNER IS NULL OR
     SEG_OWNER NOT IN ('SYS', 'SYSTEM', 'AUDSYS', 'CTXSYS', 'DVSYS', 'DBSFWUSER', 'DBSNMP', 'QSMADMIN_INTERNAL', 'LBACSYS', 'MDSYS', 'OJVMSYS', 'OLAPSYS', 'ORDDATA', 'ORDSYS', 'OUTLN', 'WMSYS', 'XDB', 'RMAN$CATALOG', 'MTSSYS', 'OML$METADATA', 'ODI_REPO_USER', 'RQSYS', 'PYQSYS'))
-    AND ((DATA_OBJ#=0 AND DATA_OBJD#=0) OR %s)`, opPredicate, strings.Join(conditions, " OR "))
+    AND ((DATA_OBJ#=0 AND DATA_OBJD#=0) OR %s) %s`, opPredicate, strings.Join(conditions, " OR "), txPredicate)
+
+	return s.conn.PrepareContext(ctx, query)
+}
+
+func (s *replicationStream) pendingAndRollbackedTransactions(ctx context.Context, startSCN, endSCN int64) (map[string]int64, map[string]struct{}, error) {
+	// we use maps for O(1) lookup
+	var excludeXIDs = make(map[string]struct{})
```

Comment:
You could make the types a bit more self-explanatory in various places by defining aliases `type XID = string` and `type SCN = int64`. That would make these ones `map[XID]struct{}` and `map[XID]SCN` for instance.